Tools/build/generate_install_details_file.py

```diff
@@ -0,0 +1,65 @@
+import argparse
+import io
+import os
+import sys
+import sysconfig
+import tomllib
+
+from typing import Any, Callable
+
+
+ConfigItem = str | int | list[str | int]
+ConfigData = dict[str, dict[str, ConfigItem]]
+
+
+def generic_info() -> ConfigData:
+    return {
+        'python': {
+            'version': sys.version.split(' ')[0],
+            'version_parts': {
+                field: getattr(sys.version_info, field)
+                for field in ('major', 'minor', 'micro', 'releaselevel', 'serial')
+            },
+            'executable': os.path.join(
+                sysconfig.get_path('scripts'),
+                os.path.basename(sys.executable),
+            ),
+            'stdlib': sysconfig.get_path('stdlib'),
+        },
+    }
+
+
+def toml_dump(fd: io.TextIOBase, data: ConfigData) -> None:
+    """**Very** basic TOML writer. It only implements what is necessary for this use-case."""
+    def toml_repr(obj: object) -> str:
+        if isinstance(obj, dict):
+            return '{ ' + ', '.join(f'{k} = {toml_repr(v)}' for k, v in obj.items()) + ' }'
+        else:
+            return repr(obj)
+
+    for section, entries in data.items():
+        print(f'[{section}]', file=fd)
+        for name, value in entries.items():
+            print(f'{name} = {toml_repr(value)}', file=fd)
+        fd.write('\n')
```

Comment:
This doesn't work on Windows, as `repr(value)` escapes backslashes and uses single quotes, which TOML interprets as [literal strings](https://toml.io/en/v1.0.0#string).

I think it would also be nice if the parser emitted dicts as tables:

```toml
[python.version_parts]
major = 3
minor = 11
micro = 5
releaselevel = 'final'
serial = 0
```

The following suggestion does both, but the fix for Windows is just changing ``repr(obj)`` to ``'{obj'}``.

```suggestion
def toml_dump(fd: io.TextIOBase, data: list[tuple[str, dict[str, ConfigItem]]]) -> None:
    """**Very** basic TOML writer. It only implements what is necessary for this use-case."""
    for section, entries in data:
        if fd.tell() != 0:  # don't emit a newline before the first section
            fd.write('\n')
        fd.write(f'[{section}]\n')
        subsections = []
        for name, value in entries.items():
            if isinstance(value, dict):
                subsections.append((f'{section}.{name}', value))
            elif isinstance(value, str):
                    fd.write(f"{name} = '{value}'\n")
            else:
                fd.write(f"{name} = {value!r}\n")

        toml_dump(fd, subsections)
```

(This also involves changing Line 57 to ``toml_dump(f, config.items())``.

Reply:
> This doesn't work on Windows, as `repr(value)` escapes backslashes and uses single quotes, which TOML interprets as [literal strings](https://toml.io/en/v1.0.0#string).

Thank you for catching this! I actually wasn't aware of the difference between basic and literal strings in TOML. I don't think we should run into any such instances here, at least for now, given the small scope, but we should try to avoid that if possible.

> I think it would also be nice if the parser emitted dicts as tables:
> 
> ```toml
> [python.version_parts]
> major = 3
> minor = 11
> micro = 5
> releaselevel = 'final'
> serial = 0
> ```

Yeah, I don't have any very strong thoughts about it, so we can go with whatever makes the implementation simpler.

> The following suggestion does both, but the fix for Windows is just changing `repr(obj)` to `'{obj'}`.

That causes other issues, like escaping quotes. I think I am just gonna make an escape table and call `str.translate`.

Reply:
The problem is more that in the previous version, e.g. the ``executable`` string would be ``'S:\\Development\\tmp\\venv\\Scripts\\python.exe'`` on Windows, which then was printed to the TOML file as ``executable = 'S:\\Development\\tmp\\venv\\Scripts\\python.exe'``, but then we hit the literal-vs-escaped string behaviour in TOML.

The PR as it stands now (with translation table) prints the following to TOML: ``executable = "S:\\\\Development\\\\tmp\\\\venv\\\\Scripts\\\\python.exe"`` -- the quotation marks are right but we've now got doubled-up backslashes!

I think that the fix is:

```diff
-    0x005C: r'\\\\',
+    0x005C: '\\\\',
```

A

Reply:
This thread makes me want to vote more strongly for either JSON or simple `key=value` with no quoting or escaping.

TOML parsers are not universal, which means a lot of tools are going to end up hacking together just enough of a parser to make things work enough of the time. Better to go with either something that's universal-enough, or something that's trivial to parse.

Reply:
I understand.

> This thread makes me want to vote more strongly for either JSON or simple `key=value` with no quoting or escaping.

IMO the simple `key=value` option wouldn't be a great option for the following reasons:
  - Being a custom format, it requires a custom parser anyway, which no matter how simple we intend it to be, it'd probably end up resulting in more user code
    - If we support different data types, the user implementation wouldn't be trivial, which I think goes against the purpose of having a "simple" mapping
    - If everything is a string, since we will still need to represent non-string data (eg. ints, bools, etc.), I think users would either end up implementing the type parsing code anyway, or it would be much more cumbersome for them to use the data, as they would likely still need to convert it to a useful type so that they can actually use the information
  - The lack of sections makes organization worse, makes it less ergonomic to have direct maps to the `sysconfig` API [^1], and makes it more cumbersome for users to use the data (eg. the `version_parts` field would need to be split in 5 different fields, on top of that, each likely with a large prefix)

That said, JSON is a reasonable choice, I just dislike how annoying it is to edit it, maybe even painful if you aren't very well familiarized with it. Not that is a big deal, it's just my personal opinion.

[^1]: I would like to have direct mappings, where possible, from this file to the data structures in the new `sysconfig` API.

> TOML parsers are not universal, which means a lot of tools are going to end up hacking together just enough of a parser to make things work enough of the time. Better to go with either something that's universal-enough, or something that's trivial to parse.

That is fair, I guess, though all major languages I am aware of have a TOML parser library, and looking at TOML's adoption rate I think that will only improve, so, realistically, I don't think this is really an issue. To be clear, I completely understand the concern, though.

Anyway, maybe I am just being a bit too opinionated here :rofl: